packages/dds/tree/src/forest/types/reader.ts

```diff
@@ -0,0 +1,25 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Serializable } from "@fluidframework/datastore-definitions";
+import { TreeKey, TreeType } from "../..";
+
+export const enum TreeReadResult {
+    NotFound = -1,
+    Pending = 0,
+    Ok = 1,
+}
+
+export interface INodeReader {
+    type: TreeType;
+    keys: Iterable<TreeKey>;
+    length(key: TreeKey): number;
+    value: undefined | Serializable;
+}
+
+export interface ITreeReader extends INodeReader {
+    push(key: TreeKey, index: number): TreeReadResult;
```

Comment:
I think "enterField" "down" or "toChild" paired with "exitField", "up" or "toParent" might be more self descriptive names.

Also do we want a way to navigate to a sibling without going up to the parent and back down? This seems important for efficient chunk traversal.

Reply:
I like down()/up().  They are equally succinct and continue to work if we expand the functionality to allow pushing an entire path and popping multiple parents:

.down("meshes", 0, "points", 0, "x")
.up(5)